src/dune_rules/scope.ml

```diff
@@ -26,158 +26,307 @@ module DB = struct
+                    | ( Deprecated_library_name (loc2, lib2)
+                      , Redirect { loc = loc1; to_ = lib1; _ } )
+                    | ( Redirect { loc = loc1; to_ = lib1; _ }
+                      , Deprecated_library_name (loc2, lib2) )
+                    | ( Deprecated_library_name (loc1, lib1)
+                      , Deprecated_library_name (loc2, lib2) ) ->
+                      if Lib_name.equal lib1 lib2 then Ok r1 else Error (loc1, loc2)
+                  in
+                  (match res with
+                   | Ok x -> Some x
+                   | Error (loc1, loc2) ->
+                     let main_message =
+                       Pp.textf "Library %s is defined twice:" (Lib_name.to_string name)
+                     in
+                     let annots =
+                       let main = User_message.make ~loc:loc2 [ main_message ] in
+                       let related =
+                         [ User_message.make ~loc:loc1 [ Pp.text "Already defined here" ]
+                         ]
+                       in
+                       User_message.Annots.singleton
+                         Compound_user_error.annot
+                         [ Compound_user_error.make ~main ~related ]
+                     in
+                     User_error.raise
+                       ~annots
+                       [ main_message
+                       ; Pp.textf "- %s" (Loc.to_file_colon_line loc1)
+                       ; Pp.textf "- %s" (Loc.to_file_colon_line loc2)
+                       ]))
+            and id_map' =
+              let id_map : Lib_info.Library_id.Set.t Lib_name.Map.t = id_map in
+              Lib_name.Map.update id_map name ~f:(fun library_ids ->
+                Some
+                  (match
+                     Option.map library_ids ~f:(fun library_ids ->
+                       Lib_info.Library_id.Set.add library_ids library_id)
+                   with
+                   | None -> Lib_info.Library_id.Set.singleton library_id
+                   | Some s -> s))
+            and library_id_map' =
+              Lib_info.Library_id.Map.add_exn library_id_map library_id r2
+            in
+            libname_map', id_map', library_id_map')
+      in
+      library_id_map, id_map
+    in
+    let resolve_library_id library_id =
+      match Lib_info.Library_id.Map.find library_id_map library_id with
+      | None -> Memo.return Lib.DB.Resolve_result.not_found
+      | Some (Redirect { loc; to_; enabled; _ }) ->
+        let+ enabled =
+          let+ toggle = Memo.Lazy.force enabled in
+          Toggle.enabled toggle
+        in
+        if enabled
+        then Lib.DB.Resolve_result.redirect_in_the_same_db (loc, to_)
+        else Lib.DB.Resolve_result.not_found
+      | Some (Found lib) -> Memo.return (Lib.DB.Resolve_result.found lib)
+      | Some (Deprecated_library_name lib) ->
+        Memo.return (Lib.DB.Resolve_result.redirect_in_the_same_db lib)
```

Comment:
Now that I think about it, this seems wrong. We should be redirecting the `new_public_name` into the public lib DB rather then the same db. This isn't an issue introduced by this PR though.